components/sync15/src/clients/engine.rs

```diff
@@ -197,43 +208,57 @@ impl<'a> Engine<'a> {
     /// For these reasons, we implement this engine directly in the `sync15`
     /// crate, and provide a specialized `sync` method instead of implementing
     /// `sync15::Store`.
-    pub fn sync(&self) -> Result<()> {
+    pub fn sync(
+        &mut self,
+        storage_client: &Sync15StorageClient,
+        global_state: &GlobalState,
+        root_sync_key: &KeyBundle,
+    ) -> Result<()> {
         log::info!("Syncing collection clients");
 
-        let coll_keys = CollectionKeys::from_encrypted_bso(
-            self.global_state.keys.clone(),
-            &self.root_sync_key,
-        )?;
+        let coll_keys =
+            CollectionKeys::from_encrypted_bso(global_state.keys.clone(), &root_sync_key)?;
         let mut coll_state = CollState {
-            config: self.global_state.config.clone(),
-            last_modified: self
-                .global_state
+            config: global_state.config.clone(),
+            last_modified: global_state
                 .collections
                 .get(COLLECTION_NAME)
                 .cloned()
                 .unwrap_or_default(),
             key: coll_keys.key_for_collection(COLLECTION_NAME).clone(),
         };
 
-        let inbound = self.fetch_incoming(&mut coll_state)?;
+        let inbound = self.fetch_incoming(&storage_client, &mut coll_state)?;
 
         let driver = Driver {
             command_processor: self.command_processor,
             interruptee: self.interruptee,
-            config: &self.global_state.config,
+            config: &global_state.config,
         };
 
-        let outgoing = driver.sync(inbound)?;
+        let mut outgoing = OutgoingChangeset::new(COLLECTION_NAME.into(), inbound.timestamp);
+        outgoing.timestamp = inbound.timestamp;
+
+        let clients = driver.sync(inbound)?;
+        let mut recent_client_list = HashMap::new();
+        for client in clients {
+            recent_client_list.insert(
```

Comment:
This should be populated from `incoming`, _not_ `outgoing`! @eoger kindly pointed out `recent_client_list` won't have any client records if we're not uploading commands.